File: app/src/main/java/com/aspirasoft/huntrek/activity/ActivitySignIn.java
package com.aspirasoft.huntrek.activity;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import com.aspirasoft.huntrek.HuntItApp;
import com.aspirasoft.huntrek.R;
import com.aspirasoft.huntrek.bo.UserManager;
import com.aspirasoft.huntrek.model.User;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.FirebaseException;
import com.google.firebase.FirebaseTooManyRequestsException;
import com.google.firebase.auth.*;
import com.google.firebase.database.*;

import java.util.concurrent.TimeUnit;

public class ActivitySignIn extends AppCompatActivity implements
        View.OnClickListener {

    private static final String KEY_VERIFY_IN_PROGRESS = "key_verify_in_progress";

    private static final int STATE_INITIALIZED = 1;
    private static final int STATE_AWAITING_RESPONSE = 2;
    private static final int STATE_CODE_SENT = 3;
    private static final int STATE_VERIFY_FAILED = 4;
    private static final int STATE_VERIFY_SUCCESS = 5;
    private static final int STATE_SIGNIN_FAILED = 6;
    private static final int STATE_SIGNIN_SUCCESS = 7;

    private FirebaseAuth mAuth;

    private boolean mVerificationInProgress = false;
    private String mVerificationId;
    private PhoneAuthProvider.ForceResendingToken mResendToken;
    private PhoneAuthProvider.OnVerificationStateChangedCallbacks mCallbacks;

    private ViewGroup mLoginViews;
    private ViewGroup mVerificationViews;
    private ViewGroup mWaitingViews;

    private EditText mPhoneNumberField;
    private EditText mVerificationField;

    private Button mStartButton;
    private Button mVerifyButton;
    private TextView mResendLink;

    private UserManager mUserManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        View decorView = getWindow().getDecorView();
        int uiOptions = View.SYSTEM_UI_FLAG_LAYOUT_STABLE | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN;
        decorView.setSystemUiVisibility(uiOptions);
        setContentView(R.layout.activity_sign_in);

        // Restore instance state
        if (savedInstanceState != null) {
            onRestoreInstanceState(savedInstanceState);
        }

        mUserManager = UserManager.getInstance();

        // Assign views
        mLoginViews = findViewById(R.id.loginViews);
        mVerificationViews = findViewById(R.id.verificationViews);
        mWaitingViews = findViewById(R.id.waitingViews);

        mPhoneNumberField = findViewById(R.id.field_phone_number);
        mVerificationField = findViewById(R.id.field_verification_code);

        mStartButton = findViewById(R.id.buttonStartVerification);
        mVerifyButton = findViewById(R.id.buttonVerifyPhone);
        mResendLink = findViewById(R.id.linkResend);

        // Assign click listeners
        mStartButton.setOnClickListener(this);
        mVerifyButton.setOnClickListener(this);
        mResendLink.setOnClickListener(this);

        // [START initialize_auth]
        mAuth = FirebaseAuth.getInstance();
        // [END initialize_auth]

        // Initialize phone auth callbacks
        // [START phone_auth_callbacks]
        mCallbacks = new PhoneAuthProvider.OnVerificationStateChangedCallbacks() {

            @Override
            public void onVerificationCompleted(PhoneAuthCredential credential) {
                // This callback will be invoked in two situations:
                // 1 - Instant verification. In some cases the phone number can be instantly
                //     verified without needing to send or enter a verification code.
                // 2 - Auto-retrieval. On some devices Google Play services can automatically
                //     detect the incoming verification SMS and perform verification without
                //     user action.
                Log.d(HuntItApp.TAG, "onVerificationCompleted:" + credential);
                // [START_EXCLUDE silent]
                mVerificationInProgress = false;
                // [END_EXCLUDE]

                // [START_EXCLUDE silent]
                // Update the UI and attempt sign in with the phone credential
                updateUI(STATE_VERIFY_SUCCESS, null, credential);
                // [END_EXCLUDE]
                signInWithPhoneAuthCredential(credential);
            }

            @Override
            public void onVerificationFailed(FirebaseException e) {
                // This callback is invoked in an invalid request for verification is made,
                // for instance if the the phone number format is not valid.
                Log.w(HuntItApp.TAG, "onVerificationFailed", e);
                // [START_EXCLUDE silent]
                mVerificationInProgress = false;
                // [END_EXCLUDE]

                if (e instanceof FirebaseAuthInvalidCredentialsException) {
                    // Invalid request
                    // [START_EXCLUDE]
                    mPhoneNumberField.setError("Invalid phone number.");
                    // [END_EXCLUDE]
                } else if (e instanceof FirebaseTooManyRequestsException) {
                    // The SMS quota for the project has been exceeded
                    // [START_EXCLUDE]
                    Snackbar.make(findViewById(android.R.id.content), "Quota exceeded.",
                            Snackbar.LENGTH_SHORT).show();
                    // [END_EXCLUDE]
                }

                // Show a message and update the UI
                // [START_EXCLUDE]
                updateUI(STATE_VERIFY_FAILED);
                // [END_EXCLUDE]
            }

            @Override
            public void onCodeSent(String verificationId,
                                   PhoneAuthProvider.ForceResendingToken token) {
                // The SMS verification code has been sent to the provided phone number, we
                // now need to ask the user to enter the code and then construct a credential
                // by combining the code with a verification ID.
                Log.d(HuntItApp.TAG, "onCodeSent:" + verificationId);

                // Save verification ID and resending token so we can use them later
                mVerificationId = verificationId;
                mResendToken = token;

                // [START_EXCLUDE]
                // Update UI
                updateUI(STATE_CODE_SENT);
                // [END_EXCLUDE]
            }
        };
        // [END phone_auth_callbacks]
    }

    @Override
    public void onStart() {
        super.onStart();
        // Check if user is signed in and update UI accordingly.
        FirebaseUser currentUser = mAuth.getCurrentUser();
        if (currentUser == null) {
            updateUI(STATE_INITIALIZED);
        } else {
            mUserManager.isRegistered(currentUser.getUid(), new UserManager.RegistrationQueryCallback() {
                @Override
                public void onResponseReceived(boolean isRegistered) {
                    if (isRegistered) {
                        updateUI(STATE_SIGNIN_SUCCESS, null, null);
                    } else {
                        updateUI(STATE_INITIALIZED);
                    }
                }
            });
        }

        // Resume phone number validation (if previously started)
        if (mVerificationInProgress && validatePhoneNumber()) {
            startPhoneNumberVerification(mPhoneNumberField.getText().toString());
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(KEY_VERIFY_IN_PROGRESS, mVerificationInProgress);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        mVerificationInProgress = savedInstanceState.getBoolean(KEY_VERIFY_IN_PROGRESS);
    }


    private void startPhoneNumberVerification(String phoneNumber) {
        updateUI(STATE_AWAITING_RESPONSE);

        PhoneAuthProvider.getInstance().verifyPhoneNumber(
                phoneNumber,            // Phone number to verify
                60,                  // Timeout duration
                TimeUnit.SECONDS,       // Unit of timeout
                this,           // Activity (for callback binding)
                mCallbacks);            // OnVerificationStateChangedCallbacks

        mVerificationInProgress = true;
    }

    private void verifyPhoneNumberWithCode(String verificationId, String code) {
        // [START verify_with_code]
        PhoneAuthCredential credential = PhoneAuthProvider.getCredential(verificationId, code);
        // [END verify_with_code]
        signInWithPhoneAuthCredential(credential);
    }

    private void resendVerificationCode(String phoneNumber,
                                        PhoneAuthProvider.ForceResendingToken token) {
        updateUI(STATE_AWAITING_RESPONSE);

        PhoneAuthProvider.getInstance().verifyPhoneNumber(
                phoneNumber,        // Phone number to verify
                60,              // Timeout duration
                TimeUnit.SECONDS,   // Unit of timeout
                this,       // Activity (for callback binding)
                mCallbacks,         // OnVerificationStateChangedCallbacks
                token);             // ForceResendingToken from callbacks
    }

    private void signInWithPhoneAuthCredential(PhoneAuthCredential credential) {
        updateUI(STATE_AWAITING_RESPONSE);

        mAuth.signInWithCredential(credential)
                .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            // Sign in success, update UI with the signed-in user's information
                            Log.d(HuntItApp.TAG, "signInWithCredential:success");

                            FirebaseUser user = task.getResult().getUser();
                            // [START_EXCLUDE]
                            updateUI(STATE_SIGNIN_SUCCESS, user, null);
                            // [END_EXCLUDE]
                        } else {
                            // Sign in failed, display a message and update the UI
                            Log.w(HuntItApp.TAG, "signInWithCredential:failure", task.getException());
                            if (task.getException() instanceof FirebaseAuthInvalidCredentialsException) {
                                // The verification code entered was invalid
                                // [START_EXCLUDE silent]
                                mVerificationField.setError("Invalid code.");
                                // [END_EXCLUDE]
                            }
                            // [START_EXCLUDE silent]
                            // Update UI
                            updateUI(STATE_SIGNIN_FAILED);
                            // [END_EXCLUDE]
                        }
                    }
                });
    }

    private void updateUI(int uiState) {
        updateUI(uiState, mAuth.getCurrentUser(), null);
    }

    private void updateUI(int uiState, FirebaseUser user, PhoneAuthCredential cred) {
        switch (uiState) {
            case STATE_INITIALIZED:
                mLoginViews.setVisibility(View.VISIBLE);
                mWaitingViews.setVisibility(View.GONE);
                mVerificationViews.setVisibility(View.GONE);
                break;
            case STATE_AWAITING_RESPONSE:
                mLoginViews.setVisibility(View.GONE);
                mWaitingViews.setVisibility(View.VISIBLE);
                mVerificationViews.setVisibility(View.GONE);
                break;
            case STATE_CODE_SENT:
                mLoginViews.setVisibility(View.GONE);
                mWaitingViews.setVisibility(View.GONE);
                mVerificationViews.setVisibility(View.VISIBLE);
                break;
            case STATE_VERIFY_FAILED:
                mLoginViews.setVisibility(View.GONE);
                mWaitingViews.setVisibility(View.GONE);
                mVerificationViews.setVisibility(View.VISIBLE);
                break;
            case STATE_VERIFY_SUCCESS:
                FirebaseUser currentUser = mAuth.getCurrentUser();
                if (currentUser != null) {
                    UserManager.getInstance().isRegistered(currentUser.getUid(),
                            new UserManager.RegistrationQueryCallback() {
                                @Override
                                public void onResponseReceived(boolean isRegistered) {
                                    if (isRegistered) {
                                        updateUI(STATE_SIGNIN_SUCCESS);
                                    } else {
                                        startActivity(new Intent(getApplicationContext(), ActivitySignUp.class));
                                        overridePendingTransition(0, 0);
                                        finish();
                                    }
                                }
                            });
                }
                break;
            case STATE_SIGNIN_FAILED:
                mLoginViews.setVisibility(View.VISIBLE);
                mWaitingViews.setVisibility(View.GONE);
                mVerificationViews.setVisibility(View.GONE);
                break;
            case STATE_SIGNIN_SUCCESS:
                // Signed in
                DatabaseReference ref = FirebaseDatabase.getInstance().getReference().child("users").child(user.getUid());
                ref.addListenerForSingleValueEvent(new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot dataSnapshot) {
                        UserManager.getInstance().signIn(dataSnapshot.getValue(User.class));
                        startActivity(new Intent(getApplicationContext(), ActivityHunt.class));
                        overridePendingTransition(0, 0);
                        finish();
                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {

                    }
                });
                break;
        }

        if (user == null) {
            // Signed out
            mLoginViews.setVisibility(mVerificationInProgress ? View.GONE : View.VISIBLE);
            mVerificationViews.setVisibility(mVerificationInProgress ? View.VISIBLE : View.GONE);
        } else {
            // Signed in
            DatabaseReference ref = FirebaseDatabase.getInstance().getReference().child("users").child(user.getUid());
            ref.addListenerForSingleValueEvent(new ValueEventListener() {
                @Override
                public void onDataChange(DataSnapshot dataSnapshot) {
                    UserManager.getInstance().signIn(dataSnapshot.getValue(User.class));
                    startActivity(new Intent(getApplicationContext(), ActivityHunt.class));
                    overridePendingTransition(0, 0);
                    finish();
                }

                @Override
                public void onCancelled(DatabaseError databaseError) {

                }
            });
        }
    }

    private boolean validatePhoneNumber() {
        String phoneNumber = mPhoneNumberField.getText().toString();
        if (TextUtils.isEmpty(phoneNumber)) {
            mPhoneNumberField.setError("Invalid phone number.");
            return false;
        }

        return true;
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.buttonStartVerification:
                if (!validatePhoneNumber()) {
                    return;
                }

                startPhoneNumberVerification(mPhoneNumberField.getText().toString());
                break;
            case R.id.buttonVerifyPhone:
                String code = mVerificationField.getText().toString();
                if (TextUtils.isEmpty(code)) {
                    mVerificationField.setError("Cannot be empty.");
                    return;
                }

                verifyPhoneNumberWithCode(mVerificationId, code);
                break;
            case R.id.linkResend:
                resendVerificationCode(mPhoneNumberField.getText().toString(), mResendToken);
                break;
        }
    }

}

File: app/src/main/java/dev/aspirasoft/huntrek/ui/activity/ActivitySignIn.kt
package dev.aspirasoft.huntrek.ui.activity

import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.FirebaseException
import com.google.firebase.FirebaseTooManyRequestsException
import com.google.firebase.auth.*
import com.google.firebase.auth.PhoneAuthProvider.ForceResendingToken
import com.google.firebase.auth.PhoneAuthProvider.OnVerificationStateChangedCallbacks
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import dev.aspirasoft.huntrek.data.repo.AuthRepository
import dev.aspirasoft.huntrek.data.source.DataSource
import dev.aspirasoft.huntrek.model.GameCharacterState
import kotlinx.android.synthetic.main.activity_sign_in.*
import java.util.concurrent.TimeUnit

class ActivitySignIn : FullScreenActivity(), View.OnClickListener {

    private var mVerificationInProgress = false
    private var mVerificationId: String? = null

    private var mResendToken: ForceResendingToken? = null

    /**
     * Callbacks for phone authentication.
     */
    private val mCallbacks: OnVerificationStateChangedCallbacks = object : OnVerificationStateChangedCallbacks() {
        /**
         * For instance verification or auto-retrieval of verification code.
         *
         * This callback will be invoked in two situations:
         * 1.   Instant verification. In some cases the phone number can be instantly
         *      verified without needing to send or enter a verification code.
         * 2.   Auto-retrieval. On some devices Google Play services can automatically
         *      detect the incoming verification SMS and perform verification without
         *      user action.
         */
        override fun onVerificationCompleted(credential: PhoneAuthCredential) {
            mVerificationInProgress = false
            updateUI(STATE_VERIFY_SUCCESS, null)
            signInWithPhoneAuthCredential(credential)
        }

        /**
         * This callback is invoked in an invalid request for verification is made,
         * for instance if the phone number format is not valid.
         */
        override fun onVerificationFailed(ex: FirebaseException) {
            mVerificationInProgress = false
            when (ex) {
                is FirebaseAuthInvalidCredentialsException -> {
                    fieldPhoneNumber.error = "Invalid phone number."
                }
                is FirebaseTooManyRequestsException -> {
                    Snackbar.make(findViewById(android.R.id.content), "Quota exceeded.", Snackbar.LENGTH_SHORT).show()
                }
                else -> Snackbar.make(findViewById(android.R.id.content),
                    ex.message ?: "Unknown error.", Snackbar.LENGTH_LONG).show()
            }

            // Show a message and update the UI
            updateUI(STATE_VERIFY_FAILED)
        }

        /**
         *  The SMS verification code has been sent to the provided phone number, we
         *  now need to ask the user to enter the code and then construct a credential
         *  by combining the code with a verification ID.
         */
        override fun onCodeSent(verificationId: String, token: ForceResendingToken) {
            // Save verification ID and resending token so we can use them later
            mVerificationId = verificationId
            mResendToken = token

            // Update UI
            updateUI(STATE_CODE_SENT)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sign_in)

        // Restore instance state
        savedInstanceState?.let { onRestoreInstanceState(it) }

        // Assign click listeners
        buttonStartVerification.setOnClickListener(this)
        buttonVerifyPhone.setOnClickListener(this)
        linkResend.setOnClickListener(this)
    }

    public override fun onStart() {
        super.onStart()
        // Check if user is signed in and update UI accordingly.
        val currentUser = AuthRepository.currentFirebaseUser
        if (currentUser == null) {
            updateUI(STATE_INITIALIZED)
        } else {
            AuthRepository.isRegistered(currentUser.uid, object : AuthRepository.RegistrationQueryCallback {
                override fun onResponseReceived(isRegistered: Boolean) {
                    if (isRegistered) {
                        updateUI(STATE_SIGN_IN_SUCCESS, null)
                    } else {
                        updateUI(STATE_INITIALIZED)
                    }
                }
            })
        }
        // Resume phone number validation (if previously started)
        if (mVerificationInProgress && validatePhoneNumber()) {
            startPhoneNumberVerification(fieldPhoneNumber!!.text.toString())
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(KEY_VERIFY_IN_PROGRESS, mVerificationInProgress)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        mVerificationInProgress = savedInstanceState.getBoolean(KEY_VERIFY_IN_PROGRESS)
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.buttonStartVerification -> {
                if (!validatePhoneNumber()) return
                startPhoneNumberVerification(fieldPhoneNumber!!.text.toString())
            }
            R.id.buttonVerifyPhone -> {
                val code = fieldVerificationCode!!.text.toString()
                if (TextUtils.isEmpty(code)) {
                    fieldVerificationCode!!.error = "Cannot be empty."
                    return
                }
                verifyPhoneNumberWithCode(mVerificationId, code)
            }
            R.id.linkResend -> resendVerificationCode(fieldPhoneNumber!!.text.toString(), mResendToken)
        }
    }

    private fun startPhoneNumberVerification(phoneNumber: String) {
        updateUI(STATE_AWAITING_RESPONSE)
        val auth = FirebaseAuth.getInstance()
        val options = PhoneAuthOptions.newBuilder(auth)
            .setPhoneNumber(phoneNumber)       // Phone number to verify
            .setTimeout(60L, TimeUnit.SECONDS) // Timeout and unit
            .setActivity(this)                 // Activity (for callback binding)
            .setCallbacks(mCallbacks)          // OnVerificationStateChangedCallbacks
            .build()
        auth.useAppLanguage()
        PhoneAuthProvider.verifyPhoneNumber(options)

        mVerificationInProgress = true
    }

    private fun verifyPhoneNumberWithCode(verificationId: String?, code: String) {
        // [START verify_with_code]
        val credential = PhoneAuthProvider.getCredential(verificationId!!, code)
        // [END verify_with_code]
        signInWithPhoneAuthCredential(credential)
    }

    private fun resendVerificationCode(phoneNumber: String, token: ForceResendingToken?) {
        updateUI(STATE_AWAITING_RESPONSE)
        PhoneAuthProvider.getInstance().verifyPhoneNumber(
            phoneNumber,  // Phone number to verify
            60,  // Timeout duration
            TimeUnit.SECONDS,  // Unit of timeout
            this,  // Activity (for callback binding)
            mCallbacks,  // OnVerificationStateChangedCallbacks
            token) // ForceResendingToken from callbacks
    }

    private fun signInWithPhoneAuthCredential(credential: PhoneAuthCredential) {
        updateUI(STATE_AWAITING_RESPONSE)
        AuthRepository.firebaseAuth.signInWithCredential(credential)
            .addOnCompleteListener(this) { task ->
                // Sign in success, update UI with the signed-in user's information
                if (task.isSuccessful) {
                    val user = task.result!!.user
                    updateUI(STATE_SIGN_IN_SUCCESS, user)
                }

                // Sign in failed, display a message and update the UI
                else {
                    // The verification code entered was invalid
                    if (task.exception is FirebaseAuthInvalidCredentialsException) {
                        fieldVerificationCode!!.error = "Invalid code."
                    }

                    // Update UI
                    updateUI(STATE_SIGN_IN_FAILED)
                }
            }
    }

    private fun updateUI(uiState: Int, user: FirebaseUser? = AuthRepository.currentFirebaseUser) {
        when (uiState) {
            STATE_INITIALIZED -> {
                loginViews!!.visibility = View.VISIBLE
                waitingViews!!.visibility = View.GONE
                verificationViews!!.visibility = View.GONE
            }
            STATE_AWAITING_RESPONSE -> {
                loginViews!!.visibility = View.GONE
                waitingViews!!.visibility = View.VISIBLE
                verificationViews!!.visibility = View.GONE
            }
            STATE_CODE_SENT -> {
                loginViews!!.visibility = View.GONE
                waitingViews!!.visibility = View.GONE
                verificationViews!!.visibility = View.VISIBLE
            }
            STATE_VERIFY_FAILED -> {
                loginViews!!.visibility = View.GONE
                waitingViews!!.visibility = View.GONE
                verificationViews!!.visibility = View.VISIBLE
            }
            STATE_VERIFY_SUCCESS -> {
                val currentUser = AuthRepository.currentFirebaseUser
                if (currentUser != null) {
                    AuthRepository.isRegistered(currentUser.uid, object : AuthRepository.RegistrationQueryCallback {
                        override fun onResponseReceived(isRegistered: Boolean) {
                            if (isRegistered) {
                                updateUI(STATE_SIGN_IN_SUCCESS)
                            } else {
                                startActivity(Intent(applicationContext, ActivitySignUp::class.java))
                                overridePendingTransition(0, 0)
                                finish()
                            }
                        }
                    })
                }
            }
            STATE_SIGN_IN_FAILED -> {
                loginViews!!.visibility = View.VISIBLE
                waitingViews!!.visibility = View.GONE
                verificationViews!!.visibility = View.GONE
            }
            STATE_SIGN_IN_SUCCESS -> {
                if (user != null) {
                    // Signed in
                    DataSource.remoteDb.child("users").child(user.uid)
                        .addListenerForSingleValueEvent(object : ValueEventListener {
                            override fun onDataChange(dataSnapshot: DataSnapshot) {
                                dataSnapshot.getValue(GameCharacterState::class.java)?.let { user ->
                                    AuthRepository.signIn(user)
                                    onSignedIn()
                                } ?: AuthRepository.signOut()
                            }

                            override fun onCancelled(databaseError: DatabaseError) {
                                AuthRepository.signOut()
                            }
                        })
                }
            }
        }
    }

    private fun validatePhoneNumber(): Boolean {
        val phoneNumber = fieldPhoneNumber.text.toString()
        if (TextUtils.isEmpty(phoneNumber)) {
            fieldPhoneNumber.error = "Invalid phone number."
            return false
        }

        return true
    }

    private fun onSignedIn() {
        startActivity(Intent(applicationContext, ActivityHunt::class.java))
        overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
        finish()
    }

    companion object {
        private const val KEY_VERIFY_IN_PROGRESS = "key_verify_in_progress"
        private const val STATE_INITIALIZED = 1
        private const val STATE_AWAITING_RESPONSE = 2
        private const val STATE_CODE_SENT = 3
        private const val STATE_VERIFY_FAILED = 4
        private const val STATE_VERIFY_SUCCESS = 5
        private const val STATE_SIGN_IN_FAILED = 6
        private const val STATE_SIGN_IN_SUCCESS = 7
    }

}